Fix: add close method

ducquangkstn opened the pull request:
Add `Close` function for service provider:
- Close the net.Listener to stop accepting new connections
- Cancel the context of the go routine for handling invididual connections:
       - Once the context is cancel
       - Send event11 to lower layer of dicom protocol to terminate the state machine. 
Note:
I will need to merge https://github.com/segmed/go-dicom/pull/3 before this MR

Commit:
Fix: add close method

Commit:
Fix: tmp

Commit:
Fix: minor fix so that we can still run the test

segmed-lam commented on serviceprovider.go:
```diff
@@ -564,8 +565,18 @@ func RunProviderForConn(conn net.Conn, params ServiceProviderParams) {
+				break handleEvtLoop
+			}
+			disp.handleEvent(event)
+		case <-ctx.Done():
```
Can you add unit test for this case?

ducquangkstn replied:
The whole library just does not have much unit test. Most of them are end to end test. So I actually don't know where to start ...

segmed-lam replied:
I'm not sure about these changes without proper unit tests, may be we should refactor the tests first?

ducquangkstn replied:
Can you check if 725737a33e3d64622de1b83f324c381d7df3b5fd is enough?

segmed-lam commented on serviceprovider.go:
```diff
@@ -574,17 +585,23 @@ func RunProviderForConn(conn net.Conn, params ServiceProviderParams) {
 		if err != nil {
 			dicomlog.Vprintf(0, "dicom.serviceProvider(%s): Accept error: %v", sp.label, err)
-			continue
+			return
```
What is this change?

ducquangkstn replied:
As explained in the discussion: net.Listener.Accept will return an error when net.Listener.Close is called.

Hence it make sense to exit the loop.

segmed-lam replied:
IMO, we should exit the loop if `errors.As(err, ErrContextCanceled)`

ducquangkstn replied:
- `listener.Accept` is not a blocking call and it does not use any kind of Context. 
- `errors.Is(err, net.ErrClosed)`, I decide to return a specific error, indicating clearly that the service is closed.
- if the error is temporary, then continue.
- Else return an error

This follows the current practice of function `http.ListenAndServer` in the built-in library. 
https://github.com/golang/go/blob/master/src/net/http/server.go#L3463

segmed-lam commented on e2e_test.go:
```diff
@@ -243,6 +243,7 @@ func TestDCMTKCStore(t *testing.T) {
 
 // Test using "getscu" command from dcmtk.
 func TestDCMTKCGet(t *testing.T) {
+	t.Skip("TODO: fix this")
```
Can you fix this?

ducquangkstn replied:
There are multiple issues:
- The test is broken before I get it.
- Cget is not common (due to security)
- CI is broken (travis does not support opensource any more)

Hence we should do it in another MR.

segmed-lam replied:
Got it, could you open a ticket for these issues?

segmed-lam reviewed (commented):
@thiensegmed Would you mind take the second review?

Commit:
Fix: improve wrapping error

Commit:
Fix: add error to server.Run()

Commit:
Fix: continue for temporary error

thiensegmed commented on fuzze2e/fuzze2e.go:
```diff
@@ -29,15 +30,16 @@ func startServer(faults netdicom.FaultInjector) net.Listener {
 			},
 		}
-
+		ctx, cancel := context.WithCancel(context.Background())
```
What would trigger the  `cancel`? 
As I see this goroutine runs forever, and the context is created inside of this goroutine @@

ducquangkstn replied:
Pls read the description:

net.Listener.Accept() will accept in-comming connections until net.Listener.Close() is called. 

When dsp.Close() is called, then Listener.Close() and net.Listener.Accept() will return an error, hence the func dsp.Run will return and this context will be cancelled. The context will populate the close of connection to other go routine that handle single connection.

thiensegmed commented on serviceprovider.go:
```diff
@@ -564,27 +567,56 @@ func RunProviderForConn(conn net.Conn, params ServiceProviderParams) {
+			}
+			disp.handleEvent(event)
+		case <-ctx.Done():
+			disp.downcallCh <- stateEvent{event: evt11}
```
Should this be `evt17`?
I read `github.com/segmed/go-netdicom@v0.0.0-20250703071137-945ff85a283e/statemachine.go:894` and it shows use `evt17` to close `upcallStreamCh`
If we don't close `upcallStreamCh` stream the loop will still be continued forever, right?

ducquangkstn replied:
It will close eventually. evt11 is to initiate closing connection. 

You can see the similar thing for serviceUser.

https://github.com/segmed/go-netdicom/blob/master/serviceuser.go#L577

thiensegmed commented on serviceprovider.go:
```diff
@@ -564,27 +567,56 @@ func RunProviderForConn(conn net.Conn, params ServiceProviderParams) {
+
+// Run listens to incoming connections, accepts them, and runs the DICOM protocol.
+// This function will blocks until error or sp.Close()
+func (sp *ServiceProvider) Run() error {
```
IMO, this func should be `Run(ctx context.Context)`
The event of cancelling should allow the outside (where this `ServiceProvider` created/called) to call too, not just only this service receive the `ErrClosed`
Then the line 593 should be `iCtx, cancel := context.Cancel(ctx)` 
Example: like we run a command to start this server:
```
cmd := cobra.Cmd{
...
  RunE(cmd, args) {
    sp := NewServiceProvider(...)
    go sp.Run(cmd.Context())
    ... do other stuff
  }
}
```
by this, if the command is canceled --> the context of command will cancel everything inside
the expected behavior is `where using this can control the terminal of this`

ducquangkstn replied:
net.Listener.Accept does not work that way.

There is no other function of net.Listener to close by context.

The only way to stop `net.Listener.Accept()` is to use `net.Listener.Close()`

thiensegmed commented:
LGTM

Commit:
Fix: update go.mod

cajund commented:
Hi there,

I know this is closed, but I was inspecting this work as I have just implemented `context.Done()` handling throughout my version of this module.

In my implementation, instead of closing the DICOM network connection with an AR-1, I added my changes to the function: `func (sp *ServiceProvider) Run()` This allows the network to continue receiving the study in the event the interrupt or kill signal is received, preventing the possibility of incomplete studies. This is a significant concern in our use case. No new network connections will be initiated after the signal is received, but the ones that have started will complete.

I wanted to share to see if you had any thoughts about this approach (I've replaced with my logger).

Best.

```
func (sp *ServiceProvider) Run() {

	var wg sync.WaitGroup
	sp.Params.Ctx.Value("wg").(*sync.WaitGroup).Add(1)

	go func() {
		for {
			conn, err := sp.listener.Accept()
			if err != nil {
				if sp.Params.Ctx.Err() != nil {
					sp.Params.Logger.LogInfo(fmt.Sprintf("DICOM Server (%s): Close requested", sp.label))
					return
				}
				sp.Params.Logger.LogError(fmt.Errorf("dicom.serviceProvider(%s): Accept error: %v", sp.label, err))
				continue
			}

			wg.Add(1)
			go func(c net.Conn) {
				defer wg.Done()
				sp.Params.Logger.LogInfo(fmt.Sprintf("dicom.serviceProvider(%s): Accepted connection %p (remote: %+v)", sp.label, conn, conn.RemoteAddr()))
				RunProviderForConn(c, sp.Params)
			}(conn)
		}
	}()

	// Shutdown signal
	<-sp.Params.Ctx.Done()

	sp.Params.Logger.LogInfo("DICOM Server Shutting down")
	err := sp.listener.Close()
	if err != nil {
		sp.Params.Logger.LogError(fmt.Errorf("Error stopping DICOM Server"))
	}
	wg.Wait() // wait for current ones to finish

	sp.Params.Ctx.Value("wg").(*sync.WaitGroup).Done()
	sp.Params.Logger.LogInfo("DICOM Server Terminating")
}

ducquangkstn commented:
> Hi there,
> 
> I know this is closed, but I was inspecting this work as I have just implemented `context.Done()` handling throughout my version of this module.
> 
> In my implementation, instead of closing the DICOM network connection with an AR-1, I added my changes to the function: `func (sp *ServiceProvider) Run()` This allows the network to continue receiving the study in the event the interrupt or kill signal is received, preventing the possibility of incomplete studies. This is a significant concern in our use case. No new network connections will be initiated after the signal is received, but the ones that have started will complete.
> 
> I wanted to share to see if you had any thoughts about this approach (I've replaced with my logger).
> 
> Best.
> 
> ```
> func (sp *ServiceProvider) Run() {
> 
> 	var wg sync.WaitGroup
> 	sp.Params.Ctx.Value("wg").(*sync.WaitGroup).Add(1)
> 
> 	go func() {
> 		for {
> 			conn, err := sp.listener.Accept()
> 			if err != nil {
> 				if sp.Params.Ctx.Err() != nil {
> 					sp.Params.Logger.LogInfo(fmt.Sprintf("DICOM Server (%s): Close requested", sp.label))
> 					return
> 				}
> 				sp.Params.Logger.LogError(fmt.Errorf("dicom.serviceProvider(%s): Accept error: %v", sp.label, err))
> 				continue
> 			}
> 
> 			wg.Add(1)
> 			go func(c net.Conn) {
> 				defer wg.Done()
> 				sp.Params.Logger.LogInfo(fmt.Sprintf("dicom.serviceProvider(%s): Accepted connection %p (remote: %+v)", sp.label, conn, conn.RemoteAddr()))
> 				RunProviderForConn(c, sp.Params)
> 			}(conn)
> 		}
> 	}()
> 
> 	// Shutdown signal
> 	<-sp.Params.Ctx.Done()
> 
> 	sp.Params.Logger.LogInfo("DICOM Server Shutting down")
> 	err := sp.listener.Close()
> 	if err != nil {
> 		sp.Params.Logger.LogError(fmt.Errorf("Error stopping DICOM Server"))
> 	}
> 	wg.Wait() // wait for current ones to finish
> 
> 	sp.Params.Ctx.Value("wg").(*sync.WaitGroup).Done()
> 	sp.Params.Logger.LogInfo("DICOM Server Terminating")
> }
> ```

Hi @cajund There is one problem with our approach: it just not the way golang `net` packge works.

The call `net.Listener.Accept` is a **blocking** method.  Which means: the loop will stop there, until:
- There is a new connection
- `net.Listener.Close` is called.

Hence you need to call net.Listener.Close from another go routine.

Ref:

https://pkg.go.dev/net#Listener

```
type Listener interface {
	// Accept waits for and returns the next connection to the listener.
	Accept() ([Conn](https://pkg.go.dev/net#Conn), [error](https://pkg.go.dev/builtin#error))

	// Close closes the listener.
	// Any blocked Accept operations will be unblocked and return errors.
	Close() [error](https://pkg.go.dev/builtin#error)

	// Addr returns the listener's network address.
	Addr() [Addr](https://pkg.go.dev/net#Addr)
}
```

cajund commented:
Thanks for the reply. and you are correct. However, the code does this. There is a go function that wraps the blocking call to Accept()

```
	go func() {
		for {
			conn, err := sp.listener.Accept()
```

So that code handles incoming connections, sends them to RunProviderForConn() in a second go routine, and cycle the loop. The outer function blocks here:

```
	// Shutdown signal
	<-sp.Params.Ctx.Done()
```

waiting for the shutdown call.

So a second go routine for Close() is not needed because the Accept() is already in one. The inner one blocks at Accept(), the outer one blocks right before the call to Close().

Most importantly is to allow any connections that have already started to complete their work so there are no incomplete studies. It worked pretty well in my test, it's not in production yet.

Thanks.

ducquangkstn commented:
> Thanks for the reply. and you are correct. However, the code does this. There is a go function that wraps the blocking call to Accept()
> 
> ```
> 	go func() {
> 		for {
> 			conn, err := sp.listener.Accept()
> ```
> 
> So that code handles incoming connections, sends them to RunProviderForConn() in a second go routine, and cycle the loop. The outer function blocks here:
> 
> ```
> 	// Shutdown signal
> 	<-sp.Params.Ctx.Done()
> ```
> 
> waiting for the shutdown call.
> 
> So a second go routine for Close() is not needed because the Accept() is already in one. The inner one blocks at Accept(), the outer one blocks right before the call to Close().
> 
> Most importantly is to allow any connections that have already started to complete their work so there are no incomplete studies. It worked pretty well in my test, it's not in production yet.
> 
> Thanks.

What if there are continous connections to your server and it never closed?

Is that violating "Don't linger" principle of DICOM protocol?

ducquangkstn commented:
> This allows the network to continue receiving the study in the event the interrupt or kill signal is received, preventing the possibility of incomplete studies

Have you tried other way to prevent incomplete studies?

In software engineering, regardless the method you use, there will be always file loss. The best way to prevent this is knowing when it happens and which studies it happens, and retry the work for those studies.

cajund commented:
> What if there are continuous connections to your server and it never closed?
> Is that violating "Don't linger" principle of DICOM protocol?

If you look at the code, as soon as the done channel is unblocked, it closes the connection - so no new connections can come in. But the ones that are already open and started, all running in their own go routine, they are allowed to finish. That's what the wait group is doing.

```
	// Shutdown signal
	<-sp.Params.Ctx.Done()

	sp.Params.Logger.LogInfo("DICOM Server Shutting down")
	err := sp.listener.Close()
	if err != nil {
		sp.Params.Logger.LogError(fmt.Errorf("Error stopping DICOM Server"))
	}
	wg.Wait() // wait for current ones to finish
```

We have seen that these connections are open to send in all of the images - and sometime these are multiple connections for the same study, but we want all of the images in, we don't want to interrupt the sending until the sender says it is complete.

> Have you tried other way to prevent incomplete studies?
>
> In software engineering, regardless the method you use, there will be always file loss. The best way to prevent this is knowing when it happens and which studies it happens, and retry the work for those studies.

We have a bit of a unique use case, where we don't have a direct relationship with those sending to us. We accept the studies and send them to someone else. So it is on us to ensure we get everything in from the network connection in the event a shutdown is detected. So we don't readily have a way to "retry" this specific step. Once the files are in, we can do lots of things - but the accepting and storage of the specific files themselves is important to us (they will be sent along the way when the software starts back up).

I guess I wanted to share what we did as I saw your approach would interrupt the sending mid-study. This may be fine for your use case. But since I have enjoyed some really good success with your improvements to this library, I wanted to at least give something back in case you found value in it.

Thanks.
